### PR TITLE
added more simplifications to trig functions

### PR DESCRIPTION
**Pull request by srajangarg**
```
			arcsin				arccos				arctan				arccsc					arcsec					arccot

sin  		x 					sqrt(1-x^2)			x/sqrt(1+x^2)		1/x 					sqrt(1-(1/x)^2)			1/(x*sqrt(1+(1/x)^2))

cos 		sqrt(1-x^2)			x 					1/sqrt(1+x^2)		sqrt(1-(1/x)^2)			1/x 					1/sqrt(1+(1/x)^2)

tan         x/sqrt(1-x^2)		sqrt(1-x^2)/x		x					1/(x*sqrt(1-(1/x)^2))	x*sqrt(1-(1/x)^2)		1/x	

csc 		1/x					1/sqrt(1-x^2)		sqrt(1+x^2)/x		x 						1/sqrt(1-(1/x)^2)		x*sqrt(1+(1/x)^2)

sec 		1/sqrt(1-x^2)		1/x					sqrt(1+x^2)			1/sqrt(1-(1/x)^2)		x  						sqrt(1+(1/x)^2)

cot  		sqrt(1-x^2)/x		x/sqrt(1-x^2)		1/x					x*sqrt(1-(1/x)^2)		1/(x*sqrt(1-(1/x)^2))	x
```
The table used for conversion

### CODE REVIEW

**Comment by srajangarg** [comment]
@Sumith1896

**Comment by isuruf** [comment]
If else if formatting should be according to the style guide

File: symengine/functions.cpp [hunk 1]
```diff
@@ -327,8 +327,37 @@ RCP<const Basic> sin(const RCP<const Basic> &arg)
     if (is_a_Number(*arg) and not static_cast<const Number &>(*arg).is_exact()) {
         return static_cast<const Number &>(*arg).get_eval().sin(*arg);
     }
-    if (is_a<ASin>(*arg)) {
-        return rcp_static_cast<const ASin>(arg)->get_arg();
+
+    RCP<const Basic> i_arg;
+    if (is_a<ASin>(*arg))
```
Comment (isuruf) [inline symengine/functions.cpp]: You can introduce a new function `is_a_TrigFunction` which should be similar to `is_a_Number` and then `i_arg` would be `rcp_static_cast<const TrigFunction>(arg)->get_arg();` or simply `static_cast<const TrigFunction &>(arg).get_arg();`

**Comment by certik** [comment]
Good job. Here is an important comment though:

I think a good rule of thumb that we've been following so far in SymEngine (as well as in SymPy) is to only do automatic simplifications (i.e. those that you cannot turn off) for things that:

1. are always valid for all complex numbers of all arguments
2. Can be done quickly (e.g. `x+x->2x` is ok, but `integral(1/(1+x^2), x) -> tan(x)` is not.)
3. the amount of symbols/operations in the final expression is not larger than the initial amount (i.e. the complexity of the symbolic tree in SymEngine) --- we should formalize exactly how to compare complexities of two symbolic trees, but usually it is obvious, e.g. `5*x` is simpler than `x+x+x+x+x`.
4. the final expression has functions that are not more complicated ---- e.g. roughly from simple ones to complicated ones: number, polynomial, sqrt, exp, elementary functions (sin/cos), special functions, hypergeometric functions, etc.

So for example:

* `arcsin(sin(x)) -> x` is not valid for all x, so can't be done
* `sin(arcsin(x)) -> x` is valid for all x, initial amount of symbols are 3 ("x", "arcsin" and "sin"), final is 1, and "x" is simpler than either sin or arcsin.
* `tan(arccsc(x)) -> 1/(x*sqrt(1-(1/x)^2))` is valid for all complex "x" (is it? I think it is, but we should double check), the final expression has sqrt, x, ^, which is simpler than tan or arccsc. But the amount of operations in the final expression is around 8 (depends how you count it) vs 3 in the initial one. So the final expression is more complicated. And so I think we should not be doing this automatically. Only when the user requests it.
* `sin(0) -> 0` is ok
* `sin(pi/2) -> 1` is ok
* `sin(pi/4) -> sqrt(2)/2` is ok ---- final expression is using simpler components (sqrt is simpler than sin), the initial expression `sin(Mul(pi, Pow(4, -1)))` has the same amount of operations as the final expression `Mul(sqrt(2), Pow(2, -1))`.
* `sin(pi/8) -> sqrt(-sqrt(2)/4 + 1/2)` is not ok, the final expression has more operations that the initial one.


The reason for the rule 3., i.e. that we should never do automatic "simplification" that creates a larger/bigger expression, like in `tan(arccsc(x))`, is so that when you have an expression, then SymEngine will never make it bigger or longer. The automatic simplification is meant to simplify expressions. Smaller expressions are faster to manipulate. E.g. let's say you want to numerically evaluate `tan(arccsc(x))` for x=`0.5`. Then it's much faster and probably also more accurate to simply evaluate `arccsc(0.5)` and then call `tan` on that, than using `1/(x*sqrt(1-(1/x)^2))` and plugging in `0.5`. And so the automatic "simplification" hurts in this case.

And so we should simply provide funcitons `trig_simplify` (perhaps calling it differently) that would take `tan(arccsc(x))` and rewrite it as `1/(x*sqrt(1-(1/x)^2))`.

Note that SymPy sometimes doesn't follow the rule 3., e.g.:
```
In [3]: sin(pi/8)
Out[3]: 
    __________
   ╱   √2   1 
  ╱  - ── + ─ 
╲╱     4    2 
```
but I think it's a mistake, because then things like `sin(pi/8).n()` are necessarily slower than they could be if no simplification was done.

**Comment by certik** [comment]
So applying this to the above table, the initial expression is always of the form `f(g(x))`, and so we should only do the simplification if the final expression is `x`, `1/x` and I think that's probably it. The next more complicated result is `sqrt(1-x^2)`, which has more operations than `cos(arcsin(x))`. So everything else except `x` and `1/x` should be done in a separate function that the user needs to call manually.

Note that things get even more complicated once you add hyperbolic functions. Consider e.g. http://www.wolframalpha.com/input/?i=sin%28asinh%28x%29%29:
```
sin(asinh(x)) -> (I/2)/(x + Sqrt[1 + x^2])^I - (I/2) (x + Sqrt[1 + x^2])^I
```
There Mathematica doesn't do this automatically either, but rather you have to do by hand:
```
TrigToExp[Sin[ArcSinh[x]]]
```
So I think we should call the function `trig_to_exp`. I think that's what it is. You write `sin(x) -> (exp(i*x)-exp(-i*x))/(2*i)`, then you express a more complicated function `sin` using less complicated functions `exp`, and `i`, `+`, `-`, `*` and `/`, but for the price of having more operations. Then you plug in let's say `pi/4` or `asinh(x)` and simplify.

**Comment by srajangarg** [comment]
@certik Thanks for the succinct explanation on the topic. I too agree that the term "simpler" is subjective, and the best way to judge is by the number of operations (evaluation complexity). I also felt that some of these "simplifications" weren't really any simpler.

I will make a separate function to simplify trigonometric functions which can later be extended to hyperbolic functions as well.

**Comment by rwst** [comment]
@srajangarg that's why I personally prefer `rewrite` over `simplify` and let the user decide how to rewrite.

**Comment by srajangarg** [comment]
@rwst What changes are you suggesting?
@isuruf What needs to be done in this now?

**Comment by certik** [comment]
I would suggest to rename `simplify_trig` (unclear what it means, especially since the operation is not simplification in the sense above) to `trig_to_exp`, similar to https://reference.wolfram.com/language/ref/TrigToExp.html. As shown above, in Mathematica that's the way you apply this operation. Later we'll make `trig_to_exp` work for more cases.

I think that's what @rwst meant as well.

**Comment by certik** [comment]
Also, add tests for all the new features that you added in this PR. Otherwise it looks good.

**Comment by srajangarg** [comment]
Why would we want to convert them into `exp` ? Isn't `sin(acos(x))` just `sqrt(1-x^2)`? Why complicate things?

**Comment by certik** [comment]
This is just the name that Mathematica uses for this. We can use a different name, perhaps `trig_to_elementary`, except that trigonometric functions are also elementary, so we need a better name.

**Comment by srajangarg** [comment]
How about `trig_to_sqrt` ?

**Comment by certik** [comment]
`trig_to_sqrt` works for all of the above simplifications. In case we add more later (perhaps for other functions), we can rename this.

**Comment by srajangarg** [comment]
@certik @isuruf @Sumith1896 Needs review, is done from my side

**Comment by certik** [comment]
I think that looks great, +1 from me. Thanks for the work @srajangarg.

**Comment by certik** [comment]
Another way to call it is `trig_to_algebraic`.

**Pull request merged**